README.rst
requests-safe
-------------

This project provides an ``HTTPAdapter`` for the Requests_ library that will
disallow connections to networks that are considered unsafe to connect to.

The primary use case is to safely be able to retrieve items from a user
provided URL without potentially requesting internal or secret resources within
ones own network.

Usage
~~~~~

It's very simple to use this library, install it as part of your project and
then the following code will protect all requests against accidentally
retrieving something from an IP address in the ranges listed below!

.. code:: python

    import requests_safe
    from requests import Session

    with Session() as s:
        requests_safe.apply(s)

        s.get("https://google.com/")

There is unfortunately no global way to apply this, so using ``requests.get()``
and friends directly will not be protected by default.

IPv4 unsafe networks
~~~~~~~~~~~~~~~~~~~~

These are the IPv4 networks that are blocked from being conneted to:

=========================      ==================
         Name                  Network (CIDR)
-------------------------      ------------------
RFC1918 (private network)      10.0.0.0/8
RFC1918                        172.16.0.0/12
RFC1918                        192.168.0.0/16
Link-Local                     169.254.0.0/16
CG-NAT address space           100.64.0.0/10
Localhost/loopback             127.0.0.0/8
Wildcard IP                    0.0.0.0/32
IETF Protocol Assignments      192.0.0.0/24
TEST-NET-1                     192.0.2.0/24
RESERVED                       192.88.99.0/24
Benchmark testing              198.18.0.0/15
TEST-NET-2                     198.51.100.0/24
TEST-NET-3                     203.0.113.0/24
IP Multicast                   224.0.0.0/4
RESERVED                       240.0.0.0/4
Limited broadcast              255.255.255.255/32
=========================      ==================

IPv6 unsafe networks
~~~~~~~~~~~~~~~~~~~~

These are the IPv6 networks that are blocked from being connected to:

=============================        ==================
         Name                        Network (CIDR)
-----------------------------        ------------------
Localhost/unspecified address        ::/128
Loopback                             ::1/128
IPv4 mapped address                  ::ffff:0:0/96
IPv4 translated addresses            ::ffff:0:0:0/96
IPv4/IPv6 translation                64:ff9b::/96
Discard prefix                       100::/64
Teredo tunneling                     2001::/32
Orchid v2 (abondoned)                2001:20::/28
Documentation                        2001:db8::/32
6to4 addressing scheme               2002::/16
ULA address space                    fc00::/7
Link-local address space             fe80::/10
Global multicast                     ff00::/8
=============================        ==================

.. _Requests: http://docs.python-requests.org/en/master/


CHANGELOG.rst
0.2 (2019-04-24)
----------------

Project is open sourced!

- No further changes from 0.1

0.1 (2019-04-23)
----------------

Internal Crunch.io only version.

- New HTTP Adapter for the Requests library.



tests/test_it.py
import json
import socket

import pytest
import requests
from mocket import Mocket
from mocket.mockhttp import Entry

import requests_safe

example_com_ips = {"2606:2800:220:1:248:1893:25c8:1946", "93.184.216.34"}


def strify_ips(lst):
    return "getaddrinfo-ips: " + ", ".join([ip for (family, ip, port) in lst])


# These are considered safe and connections should be allowed without issues.
safe_ips = [
    [
        # example.com AAAA record (considered safe)
        (socket.AF_INET6, "2606:2800:220:1:248:1893:25c8:1946", 80)
    ],
    [
        # example.com A record (considered safe)
        (socket.AF_INET, "93.184.216.34", 80)
    ],
]

# The first ip address in the mix here is disallowed, but connections should still succeed
mixed_ips = [
    [
        # IP address in documentation range, disallowed
        (socket.AF_INET6, "2001:db8::1", 80),
        # example.com AAAA record (considered safe)
        (socket.AF_INET6, "2606:2800:220:1:248:1893:25c8:1946", 80),
    ],
    [
        # IP address in documentation range, disallowed
        (socket.AF_INET6, "2001:db8::1", 80),
        # example.com A record (considered safe)
        (socket.AF_INET, "93.184.216.34", 80),
    ],
    [
        # IP address in RFC1918 range, disallowed
        (socket.AF_INET, "10.10.10.58", 80),
        # example.com A record (considered safe)
        (socket.AF_INET, "93.184.216.34", 80),
    ],
]

unsafe_ips = [
    [
        # IP address in documentation range, disallowed
        (socket.AF_INET6, "2001:db8::1", 80)
    ],
    [
        # IP address in RFC1918 range, disallowed
        (socket.AF_INET, "10.10.10.58", 80)
    ],
]


def test_apply():
    s = requests.Session()

    requests_safe.apply(s)

    assert isinstance(s.adapters["http://"], requests_safe.SafeHTTPAdapter)
    assert isinstance(s.adapters["https://"], requests_safe.SafeHTTPAdapter)

    s.close()


@pytest.fixture(params=safe_ips + mixed_ips, ids=strify_ips)
def request_succeeds(request, fake_getaddrinfo):
    for (family, ip, port) in request.param:
        fake_getaddrinfo.append(
            (family, socket.SOCK_STREAM, socket.IPPROTO_TCP, "", (ip, port))
        )

        # Set up a fake entry to return some data for all of the IP's so mocket
        # doesn't try to connect if it can't find an entry.
        mocket_entry = Entry(
            "http://example.com/info.json",
            Entry.GET,
            Entry.response_cls(body=json.dumps({"ip": ip}), status=200, headers=None),
        )

        mocket_entry.location = (ip, port)
        Mocket.register(mocket_entry)


@pytest.fixture(params=unsafe_ips, ids=strify_ips)
def request_fails(request, fake_getaddrinfo):
    for (family, ip, port) in request.param:
        fake_getaddrinfo.append(
            (family, socket.SOCK_STREAM, socket.IPPROTO_TCP, "", (ip, port))
        )


@pytest.fixture
def apply_safety_net(fake_getaddrinfo):
    fake_getaddrinfo.append(
        (
            socket.AF_INET6,
            socket.SOCK_STREAM,
            socket.IPPROTO_TCP,
            "",
            ("2606:2800:220:1:248:1893:25c8:1946", 80),
        )
    )

    # Set up a fake entry to return some data for all of the IP's so mocket
    # doesn't try to connect if it can't find an entry.
    mocket_entry = Entry(
        "http://example.com/info.json",
        Entry.GET,
        Entry.response_cls(
            body=json.dumps({"ip": "2606:2800:220:1:248:1893:25c8:1946"}),
            status=200,
            headers=None,
        ),
    )

    mocket_entry.location = ("2606:2800:220:1:248:1893:25c8:1946", 80)
    Mocket.register(mocket_entry)


def test_safe_requests(session, request_succeeds):
    ret = session.get("http://example.com/info.json")

    # This will validate that we connected to a safe ip, even if the list may
    # have contained unsafe IPs
    assert ret.json()["ip"] in example_com_ips


def test_unsafe_requests_fails(session, request_fails):
    with pytest.raises(requests.exceptions.ConnectionError) as exc:
        session.get("http://exampe.com/info.json")

    assert "getaddrinfo returns an empty list" in str(exc.value)


def test_safe_requests_ip(session, apply_safety_net):
    # Verify bare IPv6 URL works
    ret = session.get("http://[2606:2800:220:1:248:1893:25c8:1946]/info.json")

    # This will validate that we connected to a safe ip, even if the list may
    # have contained unsafe IPs
    assert ret.json()["ip"] in example_com_ips


def test_redirect_to_ip_unsafe_fails(session, fake_getaddrinfo):
    fake_getaddrinfo.append(
        (
            socket.AF_INET6,
            socket.SOCK_STREAM,
            socket.IPPROTO_TCP,
            "",
            ("2606:2800:220:1:248:1893:25c8:1946", 80),
        )
    )

    # Set up a fake entry to return some data for all of the IP's so mocket
    # doesn't try to connect if it can't find an entry.
    mocket_entry = Entry(
        "http://example.com/info.json",
        Entry.GET,
        Entry.response_cls(
            body=json.dumps({"ip": "2606:2800:220:1:248:1893:25c8:1946"}),
            status=301,
            headers={"Location": "http://10.10.10.58/other.json"},
        ),
    )

    mocket_entry.location = ("2606:2800:220:1:248:1893:25c8:1946", 80)
    Mocket.register(mocket_entry)

    mocket_entry = Entry(
        "http://10.10.10.58/other.json",
        Entry.GET,
        Entry.response_cls(
            body=json.dumps({"ip": "10.10.10.58"}), status=200, headers=None
        ),
    )

    mocket_entry.location = ("10.10.10.58", 80)
    Mocket.register(mocket_entry)

    with pytest.raises(requests.exceptions.ConnectionError) as exc:
        session.get("http://exampe.com/info.json")

    assert "getaddrinfo returns an empty list" in str(exc.value)


def test_socket_raises(session, apply_safety_net, fake_socket_raises):
    with pytest.raises(requests.exceptions.ConnectionError):
        session.get("http://example.com/info.json")


def test_socket_timeout(session, apply_safety_net, fake_socket_timeout):
    with pytest.raises(requests.exceptions.ConnectTimeout):
        session.get("http://example.com/info.json")


tests/conftest.py
import json
import socket

import mocket
import pytest
import requests
from mocket.mockhttp import Entry

import requests_safe


@pytest.fixture
def disable_socket():
    with mocket.Mocketizer():
        yield


@pytest.fixture
def fake_getaddrinfo(monkeypatch, disable_socket):
    results = []

    def _getaddrinfo(host, port, family=None, socktype=None, proto=None, flags=None):
        if host == "10.10.10.58":
            return

        for entry in results:
            yield entry

    monkeypatch.setattr(socket, "getaddrinfo", _getaddrinfo)

    yield results


@pytest.fixture
def fake_socket_raises(monkeypatch, disable_socket):
    class _socket(object):
        def __init__(self, family, socktype, proto):
            pass

        def setsockopt(self, *opts):
            pass

        def settimeout(self, timeout):
            pass

        def close(self):
            pass

        def connect(self, addr):
            raise socket.error("Always fails")

    monkeypatch.setattr(socket, "socket", _socket)


@pytest.fixture
def fake_socket_timeout(monkeypatch, disable_socket):
    def _socket(family, socktype, proto):
        raise socket.timeout("This always fails")

    monkeypatch.setattr(socket, "socket", _socket)


@pytest.fixture
def session():
    with requests.Session() as s:
        requests_safe.apply(s)

        yield s


src/requests_safe/__init__.py
"""
Provide a SafeHTTPAdapter for the requests library that can be used to no
longer allow requests to unsafe networks such as RFC1918/Link-Local and other
such address space.

This is useful for allowing a user provided URL to be provided and then safely
fetched server side without potentially leaking internal documents/secrets.
"""

import ipaddress
import socket
import sys
from socket import error as SocketError
from socket import timeout as SocketTimeout

from requests.adapters import HTTPAdapter
from requests.exceptions import Timeout
from requests.sessions import Session
from urllib3.connection import HTTPConnection, VerifiedHTTPSConnection
from urllib3.connectionpool import HTTPConnectionPool, HTTPSConnectionPool
from urllib3.exceptions import ConnectTimeoutError, NewConnectionError
from urllib3.poolmanager import PoolManager
from urllib3.util import parse_url

PY2 = sys.version_info[0] == 2

UNSAFE_NETWORKS = [
    # IPv4 unsafe networks
    # RFC1918 (private network)
    ipaddress.ip_network(u"10.0.0.0/8"),
    ipaddress.ip_network(u"172.16.0.0/12"),
    ipaddress.ip_network(u"192.168.0.0/16"),
    # Link-Local
    ipaddress.ip_network(u"169.254.0.0/16"),
    # CG-NAT address space
    ipaddress.ip_network(u"100.64.0.0/10"),
    # Localhost/loopback
    ipaddress.ip_network(u"127.0.0.0/8"),
    # Any IP in Linux
    ipaddress.ip_network(u"0.0.0.0/32"),
    # IETF Protocol Assignments
    ipaddress.ip_network(u"192.0.0.0/24"),
    # TEST-NET-1
    ipaddress.ip_network(u"192.0.2.0/24"),
    # RESERVED
    ipaddress.ip_network(u"192.88.99.0/24"),
    # Benchmark testing
    ipaddress.ip_network(u"198.18.0.0/15"),
    # TEST-NET-2
    ipaddress.ip_network(u"198.51.100.0/24"),
    # TEST-NET-3
    ipaddress.ip_network(u"203.0.113.0/24"),
    # IP Multicast
    ipaddress.ip_network(u"224.0.0.0/4"),
    # RESERVED
    ipaddress.ip_network(u"240.0.0.0/4"),
    # Limited broadcast
    ipaddress.ip_network(u"255.255.255.255/32"),
    # IPv6 unsafe networks
    # Localhost/unspecified address
    ipaddress.ip_network(u"::/128"),
    # Loopback
    ipaddress.ip_network(u"::1/128"),
    # IPv4 mapped address
    ipaddress.ip_network(u"::ffff:0:0/96"),
    # IPv4 translated addresses
    ipaddress.ip_network(u"::ffff:0:0:0/96"),
    # IPv4/IPv6 translation
    ipaddress.ip_network(u"64:ff9b::/96"),
    # Discard prefix
    ipaddress.ip_network(u"100::/64"),
    # Teredo tunneling
    ipaddress.ip_network(u"2001::/32"),
    # Orchid v2 (abondoned)
    ipaddress.ip_network(u"2001:20::/28"),
    # Documentation
    ipaddress.ip_network(u"2001:db8::/32"),
    # 6to4 addressing scheme (deprecated)
    ipaddress.ip_network(u"2002::/16"),
    # ULA address space
    ipaddress.ip_network(u"fc00::/7"),
    # Link-local address space
    ipaddress.ip_network(u"fe80::/10"),
    # Global multicast
    ipaddress.ip_network(u"ff00::/8"),
]


def ip_is_safe(ip):
    """
    Loops over all of the defined unsafe networks and checks if the IP address
    we are attempting to connect to is in one of the aforementioned unsafe
    networks.
    """

    if PY2:
        ip = ip.decode()

    ip = ipaddress.ip_address(ip)

    for network in UNSAFE_NETWORKS:
        if ip in network:
            return False

    return True


def create_connection(
    address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT, socket_options=None
):
    """
    Creates a new connection to the address (hostname generally), with a
    timeout and socket options provided.

    With one minor twist:

    It won't attempt to connect a socket to an IP address that is located in
    one of the networks defined in UNSAFE_NETWORKS, silently ignoring those IP
    addresses if returned in DNS.
    """

    host, port = address

    if host.startswith("["):  # pragma: nocover
        host = host.strip("[]")
    err = None

    for res in socket.getaddrinfo(
        host,
        port,
        0,
        socket.SOCK_STREAM,
        socket.IPPROTO_TCP,  # TCP sockets only
        (
            # We provided port number numerically
            socket.AI_NUMERICSERV
            # We don't want v4 or v6 if they are not configured
            | socket.AI_ADDRCONFIG
        ),
    ):
        af, socktype, proto, canonname, sa = res
        sock = None

        # This is the ultra-important check, only allow the connection request
        # through if the IP we are about to connect to is considered safe.

        if not ip_is_safe(sa[0]):
            continue

        try:
            sock = socket.socket(af, socktype, proto)

            # If provided, set socket level options before connecting.

            for opt in socket_options:
                sock.setsockopt(*opt)

            if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
                sock.settimeout(timeout)

            sock.connect(sa)

            return sock

        except socket.error as e:
            err = e

            if sock is not None:
                sock.close()
                sock = None

    if err is not None:
        raise err

    raise socket.error("getaddrinfo returns an empty list")


class SafeHTTPConnection(HTTPConnection):
    """
    Use the base class, except for creating a new connection, where we want to
    use our create_connection that verifies the IP we are connecting to is safe
    to connect to.
    """

    def _new_conn(self):
        """
        Override a non-public member which is used to create the connection,
        yes, implementation detail, caveat emptor
        """

        extra_kw = {}

        if self.socket_options:
            extra_kw["socket_options"] = self.socket_options

        try:
            conn = create_connection(
                (self.host.rstrip("."), self.port), self.timeout, **extra_kw
            )

        except SocketTimeout:
            raise ConnectTimeoutError(
                self,
                "Connection to %s timed out. (connect timeout=%s)"
                % (self.host, self.timeout),
            )

        except SocketError as e:
            raise NewConnectionError(
                self, "Failed to establish a new connection: %s" % e
            )

        return conn


class SafeHTTPSConnection(SafeHTTPConnection, VerifiedHTTPSConnection):
    """
    Use our SafeHTTPConnection with the urllib3 provided
    VerifiedHTTPSConnection. urllib3 allows for switching between two different
    types of HTTPSConnection, but we care about safety so we only ever want to
    use the VerifiedHTTPSConnection
    """

    pass


class SafeHTTPConnectionPool(HTTPConnectionPool):
    """
    Simple override of the ConnectionCls used by HTTPConnectionPool
    """

    ConnectionCls = SafeHTTPConnection


class SafeHTTPSConnectionPool(HTTPSConnectionPool):
    """
    Simple override of the ConnectionCls used by HTTPSConnectionPool
    """

    ConnectionCls = SafeHTTPSConnection


class SafePoolManager(PoolManager):
    """
    A urllib3 PoolManager that overrides the default pool classes with ones
    that know how to verify the IP address to be connected to is considered
    safe
    """

    def __init__(self, *args, **kwargs):
        super(SafePoolManager, self).__init__(*args, **kwargs)

        self.pool_classes_by_scheme = {
            "http": SafeHTTPConnectionPool,
            "https": SafeHTTPSConnectionPool,
        }


class SafeHTTPAdapter(HTTPAdapter):
    def init_poolmanager(self, connections, maxsize, block=False, **pool_kwargs):
        """
        Initialize the pool manager using the SafePoolManager which will use
        our SafeHTTPConnection/SafeHTTPSConnection instead of the default.
        """

        # save these values for pickling (not really related to the
        # initialization of the pool manager, but copied from
        # requests.adapter.HTTPAdapter)
        self._pool_connections = connections
        self._pool_maxsize = maxsize
        self._pool_block = block

        self.poolmanager = SafePoolManager(
            num_pools=connections,
            maxsize=maxsize,
            block=block,
            strict=True,
            **pool_kwargs
        )


def apply(session):
    adapter = SafeHTTPAdapter()
    session.mount("http://", adapter)
    session.mount("https://", adapter)
